src/util/range.h

```diff
@@ -265,6 +265,88 @@ struct concat_iteratort
   second_iteratort second_begin;
 };
 
+/// Zip two ranges to make a range of pairs.
+/// On increment, both iterators are incremented.
+/// Ends when the two ranges reach their ends, raise an exception if one ends
+/// before the other.
+template <typename first_iteratort, typename second_iteratort>
+struct zip_iteratort
+{
+public:
+  using difference_type = typename first_iteratort::difference_type;
+  using value_type = std::pair<
+    typename first_iteratort::value_type,
+    typename second_iteratort::value_type>;
+  using pointer = value_type *;
+  using reference = value_type &;
+  using iterator_category = std::forward_iterator_tag;
+
+  bool operator==(const zip_iteratort &other) const
+  {
+    return first_begin == other.first_begin && first_end == other.first_end &&
+           second_begin == other.second_begin && second_end == other.second_end;
+  }
+
+  bool operator!=(const zip_iteratort &other) const
+  {
+    return !(*this == other);
+  }
+
+  /// Preincrement operator
+  zip_iteratort &operator++()
+  {
+    ++first_begin;
+    ++second_begin;
+    INVARIANT(
+      (first_begin == first_end) == (second_begin == second_end),
+      "Zipped ranges should have the same size");
+    current = first_begin != first_end && second_begin != second_end
+                ? util_make_unique<value_type>(
+                    std::make_pair(*first_begin, *second_begin))
```

Comment:
I think I would have been trying to define a `zip_with`, for which I then supplied `make_pair` or something similar as the function parameter in order to define `zip`. But I am happy for that to be a later refinement if we find a use for `zip_with`.

Reply:
`std::make_pair` is handy, where we don't have the full type of the `std::pair` to hand, because it infers the type of the pair from the types of the arguments. I think that we have the full type of the `pair` specified here in `value_type`, which may mean that this call to `std::make_pair` is not needed.